Add guide on Migrating from WordPress (#43)
humanmade/platform-documentation: other-docs/guides/migrating-from-wordpress.md

# Migrating a WordPress Codebase

> This guide covers how to migrate a typical WordPress project to HM Platform.

Reviewer: Should this be "typical legacy HM Platform project" or something like that? The guide is specific to our projects.

Author: Yeah the idea was to try keep it _just_ generic enough to call it a migrating WordPress guide that is potentially for all WP projects.

> ### Setup Composer

Reviewer 2: Incorrect header levels throughout.

> If the project does not already have a `composer.json` in the root of the repository, you should run `composer init` from the project root. This will prompt for any fields. For now, don’t bother adding any dependancies. If you already have a `composer.json` move to Remove Dependencies.

Reviewer: ```suggestion
If the project does not already have a `composer.json` in the root of the repository, you should run `composer init` from the project root. This will prompt you fill out the required fields. For now, don’t bother adding any dependencies. If you already have a `composer.json` move to Remove Dependencies.
```

Reviewer: Guess we don't have header anchors to link to yet. That'd be a good issue to add

Reviewer 2: Handled this as part of #48

### Remove Dependencies

> We’ll need to remove all the dependencies that are part of HM Platform, as those will be installed by HM Platform itself. Firstly, remove the any submodules that are included in HM Platform. For example, `wordpress` and `hm-platform` submodules. To correctly remove a submodule:

Reviewer: ```suggestion
We’ll need to remove all the dependencies that are part of HM Platform, as those will be installed by HM Platform itself. Firstly, remove any submodules that are included in HM Platform. For example, `wordpress` and `hm-platform` submodules. To correctly remove a submodule:
```

1. `git submodule deinit -f wordpress`
2. `git rm -rf wordpress`
3. `rm -rf .git/modules/wordpress`

> Do the same for `hm-platform` if you have this project dependency. HM Platform also includes the following WordPress plugins, so if you project uses git submodules or uses composer to include them, you should remove them from your project (unless you plan to disable the HM Platform module that makes use of the plugin, and continue with the plugin directly):

Reviewer: ```suggestion
Do the same for `hm-platform` if you have this project dependency. HM Platform also includes the following WordPress plugins, so if your project uses git submodules or composer to include them, you should remove them from your project (unless you plan to disable the HM Platform module that makes use of the plugin, and continue with the plugin directly):
```

Reviewer: The part in brackets doesn't really have a use case just yet and maybe adds some confusion about when you might do that. I think it could be left out of this guide myself, or maybe mentioned as a footnote with examples of when you might want to use a plugin directly over a platform module.

Also got me thinking the package list in the core module should maybe reflect which modules are active in this case as a plugin present there won't be installed to the content directory.

Author: Created https://github.com/humanmade/platform-cms-installer/issues/3 to track that

- `smart-media`
- `gaussholder`
- `aws-rekognition`
- `two-factor`
- `stream`
- `wp-simple-saml`
- `delegated-oauth`
- `elasticpress`
- `hm-redirects`
- `msm-sitemap`
- `wp-seo`
- `amp`
- `facebook-instant-articles-wp`
- `meta-tags`
- `query-monitor`
- `hm-gtm`
- `workflows`
- `wp-user-signups`

### Move wp-config.php

Installing HM Platform will replace your `wp-config.php` so you should back it up if you have application level configuration in your `wp-config.php`.

### Installing HM Platform

Now you have your composer project up and running, it’s time to add HM Platform to the project. Do so by running `composer require humanmade/platform`.

> Once HM Platform has been installed, you should see the `wordpress` directory back in the project root, a new `wp-config.php` and `index.php`. You should now also see these 3 paths added to your `.gitignore`.

Reviewer 2: They won't be added to an existing one I think?

Author: Ah good point, updated

### Restore site configuration

If you have any custom configuration in your old `wp-config.php` (such as custom PHP constants, etc. You will need to put them into a new file outside of `wp-config.php` (as changes to this file are not allowed). It’s possible your project already has the configuration split out into a `.config` directory, but if not, create a file at `.config/load.php`.

> Only copy across constants that your custom code actually needs. There should be no database constants, or any other WordPress type constants. If you have these already in your `.config` directory, you should delete those.

Reviewer: "WordPress type constants" is a bit ambiguous. Maybe "core WordPress constants relating to multisite" ?

Author: These weren't intended to be MS only, e.g. WP_SITEURL

To make `.config/load.php` actually be included and executed, you have to add it as a Composer autoload file. To do so, add an `autoload` section to your `composer.json` if you don’t already have one, and add `.config/load.php` as an autoload files.

```json
	"autoload": {
		"files": [
			".config/load.php"
		]
	},
```

You’ll need to run `composer dump-autoload` after doing this to make sure it’s actually loaded.

### Migrate $hm_platform options (when applicable)

In your old `wp-config.php` you’ll see there is a `global $hm_platform` that sets options for hm-platform. Only migrate anything that you specifically need to, as most likely HM Platform will have better defaults. In rare cases though, things will be disabled for good reason. In HM Platform, most of the same settings are supported, but it’s now done via the `composer.json` for configuration (as is all HM Platform configuration). You should be familiar with [HM Platform configuration](docs://getting-started/configuration.md) before continuing. `$hm_platform` options should go in the `platform.modules.cloud` section of the `extra` block in the `composer.json`.

```json

> "extra": {

Reviewer: This might get rendered showing errors, github itself at least highlights syntax errors

```
composer install --no-dev
```

### Setup the local server

Assuming your project uses Chassis for local development, we’ll be removing the local Chassis install, and installing the HM Platform module. If you have a setup script (such as `.bin/setup.sh`) you should remove any Chassis setup / installation steps.
Once you have cleaned out Chassis, install the `humanmade/platform-local-chassis` composer package as a dev dependency.

```
composer require --dev humanmade/platform-local-chassis
```

Once completed, install and start your local server with `composer chassis init` and then `composer chassis start`. You should now be able to navigate to http://platform.local to see the site!

### Migrating email sending domain

It's quite possible your project specifies the wp mail sending domain via the `wp_mail_from` hook. This can now be specified as setting in the `composer.json`'s `extra.platform.modules.cloud.email.email-from-address` setting:

```json

> "modules": {

Reviewer: again maybe make this valid JSON

    "cloud": {
        "email-from-address": "webmaster@mydomainname.com"
    }
}
```

### Optionally disable HM Platform branding

> As this guide is for migrating a non-HM Platform project to use HM Platform, it's possible the client relationship and understanding does warrant changing anything visible or user-facing. If you are sure this is an "under the hood" migration, and the client has not been on-boarded with HM Platform as a brand, you can disable the branding via the `platform.cms.branding` setting:

Reviewer: ```suggestion
As this guide is for migrating a non-HM Platform project to use HM Platform, it's possible the client relationship and understanding does warrant changing anything visible or user-facing. If you are sure this is an "under the hood" migration, and the client has not been on-boarded with HM Platform as a brand, you can disable the branding via the `platform.modules.cms.branding` setting:
```

```json
"modules": {
    "cms": {
        "branding": false
    }
}
```

### Optionally disable HM Platform features

> There are some features of HM Platform that are user-facing and default-on that you might want to audit. For example, image [lazy loading](docs://media/lazy-loading.md) via Guassholder is on by default. Smart Media with Cropping UI is enabled by default. You should consult the HM Platform documentation for behavior of specific modules. Again, unless there is specific reason to disable feature and modules, we recommend keeping them on.

Reviewer: ```suggestion
There are some features of HM Platform that are user-facing and default-on that you might want to audit. For example, image [lazy loading](docs://media/lazy-loading.md) via Guassholder is on by default. Smart Media with Cropping UI is enabled by default. You should consult the HM Platform documentation for the behaviour of specific modules. Again, unless there is specific reason to disable feature and modules, we recommend keeping them on.
```

> Any module can be disabled by setting its `disabled` setting to `false`:

Reviewer: ```suggestion
Any module can be disabled by setting its `enabled` setting to `false`:
```

```json
"extra": {
	"platform": {
		"modules": {

> 			"seo": false

Reviewer: 🤔 I don't think we support this syntax
```suggestion
			"seo": {
				"enabled": false
			}
```

		}
	}
}
```

### Deploying to Cloud

> The first time HM Platform is deployed, depending on the exact configuration, there may be tasks to perform on deploy. HM Platform is always-on WordPress multisite, as such any sites that are not on installed as Multisite already, will need converting via the `multisite-convert` WP CLI command.

Reviewer: ```suggestion
The first time HM Platform is deployed, depending on the exact configuration, there may be tasks to perform on deploy. HM Platform is always configured to be a WordPress multisite, as such any sites that are not on installed as Multisite already, will need converting via the `multisite-convert` WP CLI command.
```
